Codegen error on list of bools

```
>>> def f(): return [True]
... 
>>> ff = jit(nopython=True)(f)
>>> ff()
Traceback (most recent call last):
  File "/home/antoine/numba/numba/lowering.py", line 173, in lower_block
    self.lower_inst(inst)
  File "/home/antoine/numba/numba/lowering.py", line 215, in lower_inst
    val = self.lower_assign(ty, inst)
  File "/home/antoine/numba/numba/lowering.py", line 371, in lower_assign
    return self.lower_expr(ty, value)
  File "/home/antoine/numba/numba/lowering.py", line 733, in lower_expr
    return self.context.build_list(self.builder, resty, castvals)
  File "/home/antoine/numba/numba/targets/cpu.py", line 111, in build_list
    return listobj.build_list(self, builder, list_type, items)
  File "/home/antoine/numba/numba/targets/listobj.py", line 301, in build_list
    inst = ListInstance.allocate(context, builder, list_type, nitems)
  File "/home/antoine/numba/numba/targets/listobj.py", line 190, in allocate
    self._payload.allocated = nitems
  File "/home/antoine/numba/numba/targets/listobj.py", line 154, in _payload
    return get_list_payload(self._context, self._builder, self._ty, self._list)
  File "/home/antoine/numba/numba/targets/listobj.py", line 44, in get_list_payload
    return make_payload_cls(list_type)(context, builder, ref=payload)
  File "/home/antoine/numba/numba/cgutils.py", line 95, in __init__
    % (self._be_type.as_pointer(), ref.type))
AssertionError: bad ref type: expected {i64, i64, i1}*, got {i64, i64, i8}*

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/antoine/numba/numba/dispatcher.py", line 162, in _compile_for_args
    return self.compile(sig)
  File "/home/antoine/numba/numba/dispatcher.py", line 327, in compile
    flags=flags, locals=self.locals)
  File "/home/antoine/numba/numba/compiler.py", line 594, in compile_extra
    return pipeline.compile_extra(func)
  File "/home/antoine/numba/numba/compiler.py", line 317, in compile_extra
    return self.compile_bytecode(bc, func_attr=self.func_attr)
  File "/home/antoine/numba/numba/compiler.py", line 326, in compile_bytecode
    return self._compile_bytecode()
  File "/home/antoine/numba/numba/compiler.py", line 581, in _compile_bytecode
    return pm.run(self.status)
  File "/home/antoine/numba/numba/compiler.py", line 209, in run
    raise patched_exception
  File "/home/antoine/numba/numba/compiler.py", line 201, in run
    res = stage()
  File "/home/antoine/numba/numba/compiler.py", line 537, in stage_nopython_backend
    return self._backend(lowerfn, objectmode=False)
  File "/home/antoine/numba/numba/compiler.py", line 492, in _backend
    lowered = lowerfn()
  File "/home/antoine/numba/numba/compiler.py", line 483, in backend_nopython_mode
    self.flags)
  File "/home/antoine/numba/numba/compiler.py", line 723, in native_lowering_stage
    lower.lower()
  File "/home/antoine/numba/numba/lowering.py", line 100, in lower
    self.lower_normal_function(self.fndesc)
  File "/home/antoine/numba/numba/lowering.py", line 135, in lower_normal_function
    entry_block_tail = self.lower_function_body()
  File "/home/antoine/numba/numba/lowering.py", line 160, in lower_function_body
    self.lower_block(block)
  File "/home/antoine/numba/numba/lowering.py", line 178, in lower_block
    raise LoweringError(msg, inst.loc)
numba.errors.LoweringError: Failed at nopython (nopython mode backend)
Internal error:
AssertionError: bad ref type: expected {i64, i64, i1}*, got {i64, i64, i8}*
File "<stdin>", line 1
```

--- a/numba/cgutils.py
+++ b/numba/cgutils.py
@@ -52,24 +52,29 @@ def make_bytearray(buf):
     n = len(b)
     return ir.Constant(ir.ArrayType(ir.IntType(8), n), b)
 
+
 _struct_proxy_cache = {}
 
-
-def create_struct_proxy(fe_type):
+def create_struct_proxy(fe_type, kind='value'):
     """
     Returns a specialized StructProxy subclass for the given fe_type.
     """
-    res = _struct_proxy_cache.get(fe_type)
+    cache_key = (fe_type, kind)
+    res = _struct_proxy_cache.get(cache_key)
     if res is None:
-        clsname = StructProxy.__name__ + '_' + str(fe_type)
-        bases = (StructProxy,)
+        base = {'value': ValueStructProxy,
+                'data': DataStructProxy,
+                }[kind]
+        clsname = base.__name__ + '_' + str(fe_type)
+        bases = (base,)
         clsmembers = dict(_fe_type=fe_type)
         res = type(clsname, bases, clsmembers)
-        _struct_proxy_cache[fe_type] = res
+
+        _struct_proxy_cache[cache_key] = res
     return res
 
 
-class StructProxy(object):
+class _StructProxy(object):
     """
     Creates a `Structure` like interface that is constructed with information
     from DataModel instance.  FE type must have a data model that is a
@@ -81,13 +86,12 @@ class StructProxy(object):
     def __init__(self, context, builder, value=None, ref=None):
         from numba import datamodel   # Avoid circular import
         self._context = context
-        self._dmm = self._context.data_model_manager
-        self._datamodel = self._dmm[self._fe_type]
+        self._datamodel = self._context.data_model_manager[self._fe_type]
         if not isinstance(self._datamodel, datamodel.StructModel):
             raise TypeError("Not a structure model: {0}".format(self._datamodel))
         self._builder = builder
 
-        self._be_type = self._datamodel.get_value_type()
+        self._be_type = self._get_be_type(self._datamodel)
         assert not is_pointer(self._be_type)
 
         if ref is not None:
@@ -100,6 +104,15 @@ class StructProxy(object):
             self._value = alloca_once(self._builder, self._be_type, zfill=True)
             if value is not None:
                 self._builder.store(value, self._value)
+
+    def _get_be_type(self, datamodel):
+        raise NotImplementedError
+
+    def _cast_member_to_value(self, index, val):
+        raise NotImplementedError
+
+    def _cast_member_from_value(self, index, val):
+        raise NotImplementedError
 
     def _get_ptr_by_index(self, index):
         return gep_inbounds(self._builder, self._value, 0, index)
@@ -122,15 +135,15 @@ class StructProxy(object):
         Store the LLVM *value* into the named *field*.
         """
         if field.startswith('_'):
-            return super(StructProxy, self).__setattr__(field, value)
+            return super(_StructProxy, self).__setattr__(field, value)
         self[self._datamodel.get_field_position(field)] = value
 
     def __getitem__(self, index):
         """
         Load the LLVM value of the field at *index*.
         """
-
-        return self._builder.load(self._get_ptr_by_index(index))
+        member_val = self._builder.load(self._get_ptr_by_index(index))
+        return self._cast_member_to_value(index, member_val)
 
     def __setitem__(self, index, value):
         """
@@ -151,7 +164,8 @@ class StructProxy(object):
                                 "{self._datamodel}".format(value=value,
                                                            ptr=ptr,
                                                            self=self))
-        self._builder.store(value, ptr)
+        member_val = self._cast_member_from_value(index, value)
+        self._builder.store(member_val, ptr)
 
     def __len__(self):
         """
@@ -176,6 +190,37 @@ class StructProxy(object):
         assert not is_pointer(value.type)
         assert value.type == self._type, (value.type, self._type)
         self._builder.store(value, self._value)
+
+
+class ValueStructProxy(_StructProxy):
+    """
+    Create a StructProxy suitable for accessing regular values
+    (e.g. LLVM values or alloca slots).
+    """
+    def _get_be_type(self, datamodel):
+        return datamodel.get_value_type()
+
+    def _cast_member_to_value(self, index, val):
+        return val
+
+    def _cast_member_from_value(self, index, val):
+        return val
+
+
+class DataStructProxy(_StructProxy):
+    """
+    Create a StructProxy suitable for accessing data persisted in memory.
+    """
+    def _get_be_type(self, datamodel):
+        return datamodel.get_data_type()
+
+    def _cast_member_to_value(self, index, val):
+        model = self._datamodel.get_model(index)
+        return model.from_data(self._builder, val)
+
+    def _cast_member_from_value(self, index, val):
+        model = self._datamodel.get_model(index)
+        return model.as_data(self._builder, val)
 
 
 class Structure(object):

--- a/numba/datamodel/models.py
+++ b/numba/datamodel/models.py
@@ -554,13 +554,20 @@ class StructModel(CompositeModel):
         ----
         pos: int or str
             field index or field name
-
         """
         if isinstance(pos, str):
             pos = self.get_field_position(pos)
         return self._members[pos]
 
     def get_model(self, pos):
+        """
+        Get the datamodel of a field given the position or the fieldname.
+
+        Args
+        ----
+        pos: int or str
+            field index or field name
+        """
         return self._models[pos]
 
     def traverse(self, builder, value):

--- a/numba/datamodel/models.py
+++ b/numba/datamodel/models.py
@@ -560,6 +560,9 @@ class StructModel(CompositeModel):
             pos = self.get_field_position(pos)
         return self._members[pos]
 
+    def get_model(self, pos):
+        return self._models[pos]
+
     def traverse(self, builder, value):
         out = [(self.get_type(k), self.get(builder, value, k))
                 for k in self._fields]

--- a/numba/targets/arrayobj.py
+++ b/numba/targets/arrayobj.py
@@ -16,9 +16,8 @@ from numba.numpy_support import as_dtype
 from numba.numpy_support import version as numpy_version
 from numba.targets.imputils import (builtin, builtin_attr, implement,
                                     impl_attribute, impl_attribute_generic,
-                                    iternext_impl, struct_factory,
-                                    impl_ret_borrowed, impl_ret_new_ref,
-                                    impl_ret_untracked)
+                                    iternext_impl, impl_ret_borrowed,
+                                    impl_ret_new_ref, impl_ret_untracked)
 from numba.typing import signature
 from . import quicksort, slicing
 
@@ -182,7 +181,6 @@ def update_array_info(aryty, array):
                                           get_itemsize(context, aryty))
 
 
-@struct_factory(types.ArrayIterator)
 def make_arrayiter_cls(iterator_type):
     """
     Return the Structure representation of the given *iterator_type* (an
@@ -1366,7 +1364,6 @@ def array_is(context, builder, sig, args):
 #-------------------------------------------------------------------------------
 # builtin `numpy.flat` implementation
 
-@struct_factory(types.NumpyFlatType)
 def make_array_flat_cls(flatiterty):
     """
     Return the Structure representation of the given *flatiterty* (an
@@ -1375,7 +1372,6 @@ def make_array_flat_cls(flatiterty):
     return _make_flattening_iter_cls(flatiterty, 'flat')
 
 
-@struct_factory(types.NumpyNdEnumerateType)
 def make_array_ndenumerate_cls(nditerty):
     """
     Return the Structure representation of the given *nditerty* (an
@@ -1415,7 +1411,6 @@ def _increment_indices_array(context, builder, arrty, arr, indices, end_flag=Non
     _increment_indices(context, builder, arrty.ndim, shape, indices, end_flag)
 
 
-@struct_factory(types.NumpyNdIndexType)
 def make_ndindex_cls(nditerty):
     """
     Return the Structure representation of the given *nditerty* (an

--- a/numba/targets/base.py
+++ b/numba/targets/base.py
@@ -15,7 +15,6 @@ from numba import _dynfunc, _helperlib
 from numba.pythonapi import PythonAPI
 from numba.targets.imputils import (user_function, user_generator,
                                     builtin_registry, impl_attribute,
-                                    type_registry,
                                     impl_ret_borrowed)
 from . import (
     arrayobj, builtins, iterators, rangeobj, optional, slicing, tupleobj)
@@ -296,13 +295,6 @@ class BaseContext(object):
         The return value is a llvmlite.ir.Type object, or None if the type
         is an opaque pointer (???).
         """
-        try:
-            fac = type_registry.match(ty)
-        except KeyError:
-            pass
-        else:
-            return fac(self, ty)
-
         return self.data_model_manager[ty].get_data_type()
 
     def get_value_type(self, ty):

--- a/numba/targets/builtins.py
+++ b/numba/targets/builtins.py
@@ -8,7 +8,7 @@ from llvmlite.llvmpy.core import Type, Constant
 import llvmlite.llvmpy.core as lc
 
 from .imputils import (builtin, builtin_attr, implement, impl_attribute,
-                       impl_attribute_generic, iternext_impl, struct_factory,
+                       impl_attribute_generic, iternext_impl,
                        impl_ret_borrowed, impl_ret_untracked)
 from . import optional
 from .. import typing, types, cgutils, utils, intrinsics

--- a/numba/targets/imputils.py
+++ b/numba/targets/imputils.py
@@ -236,37 +236,6 @@ builtin = builtin_registry.register
 builtin_attr = builtin_registry.register_attr
 
 
-class _StructRegistry(object):
-    """
-    A registry of factories of cgutils.Structure classes.
-    """
-
-    def __init__(self):
-        self.impls = {}
-
-    def register(self, type_class):
-        """
-        Register a Structure factory function for the given *type_class*
-        (i.e. a subclass of numba.types.Type).
-        """
-        assert issubclass(type_class, types.Type)
-        def decorator(func):
-            self.impls[type_class] = func
-            return func
-        return decorator
-
-    def match(self, typ):
-        """
-        Return the Structure factory function for the given Numba type
-        instance *typ*.
-        """
-        return self.impls[typ.__class__]
-
-
-struct_registry = _StructRegistry()
-struct_factory = struct_registry.register
-
-
 def impl_ret_new_ref(ctx, builder, retty, ret):
     """
     The implementation returns a new reference.

--- a/numba/targets/imputils.py
+++ b/numba/targets/imputils.py
@@ -267,34 +267,6 @@ struct_registry = _StructRegistry()
 struct_factory = struct_registry.register
 
 
-class _TypeRegistry(object):
-
-    def __init__(self):
-        self.factories = {}
-
-    def register(self, type_class):
-        """
-        Register a LLVM type factory function for the given *type_class*
-        (i.e. a subclass of numba.types.Type).
-        """
-        assert issubclass(type_class, types.Type)
-        def decorator(func):
-            self.factories[type_class] = func
-            return func
-        return decorator
-
-    def match(self, typ):
-        """
-        Return the LLVM type factory function for the given Numba type
-        instance *typ*.
-        """
-        return self.factories[typ.__class__]
-
-
-type_registry = _TypeRegistry()
-type_factory = type_registry.register
-
-
 def impl_ret_new_ref(ctx, builder, retty, ret):
     """
     The implementation returns a new reference.

--- a/numba/targets/iterators.py
+++ b/numba/targets/iterators.py
@@ -5,7 +5,7 @@ Implementation of various iterable and iterator types.
 from numba import types, cgutils
 from numba.targets.imputils import (
     builtin, implement, iternext_impl, call_iternext, call_getiter,
-    struct_factory, impl_ret_borrowed, impl_ret_new_ref)
+    impl_ret_borrowed, impl_ret_new_ref)
 
 
 @builtin
@@ -17,7 +17,6 @@ def iterator_getiter(context, builder, sig, args):
 #-------------------------------------------------------------------------------
 # builtin `enumerate` implementation
 
-@struct_factory(types.EnumerateType)
 def make_enumerate_cls(enum_type):
     """
     Return the Structure representation of the given *enum_type* (an
@@ -80,7 +79,6 @@ def iternext_enumerate(context, builder, sig, args, result):
 #-------------------------------------------------------------------------------
 # builtin `zip` implementation
 
-@struct_factory(types.ZipType)
 def make_zip_cls(zip_type):
     """
     Return the Structure representation of the given *zip_type* (an

--- a/numba/targets/listobj.py
+++ b/numba/targets/listobj.py
@@ -30,7 +30,8 @@ def make_payload_cls(list_type):
     Return the Structure representation of the given *list_type*'s payload
     (an instance of types.List).
     """
-    return cgutils.create_struct_proxy(types.ListPayload(list_type))
+    return cgutils.create_struct_proxy(types.ListPayload(list_type),
+                                       kind='data')
 
 
 def get_list_payload(context, builder, list_type, value):
@@ -75,16 +76,19 @@ class _ListPayloadMixin(object):
 
     def getitem(self, idx):
         ptr = self._gep(idx)
-        return self._builder.load(ptr)
+        data_item = self._builder.load(ptr)
+        return self._datamodel.from_data(self._builder, data_item)
 
     def setitem(self, idx, val):
         ptr = self._gep(idx)
-        self._builder.store(val, ptr)
+        data_item = self._datamodel.as_data(self._builder, val)
+        self._builder.store(data_item, ptr)
 
     def inititem(self, idx, val):
         ptr = self._gep(idx)
-        self._builder.store(val, ptr)
-    
+        data_item = self._datamodel.as_data(self._builder, val)
+        self._builder.store(data_item, ptr)
+
     def fix_index(self, idx):
         """
         Fix negative indices by adding the size to them.  Positive
@@ -147,6 +151,7 @@ class ListInstance(_ListPayloadMixin):
         self._ty = list_type
         self._list = make_list_cls(list_type)(context, builder, list_val)
         self._itemsize = get_itemsize(context, list_type)
+        self._datamodel = context.data_model_manager[list_type.dtype]
 
     @property
     def _payload(self):
@@ -261,6 +266,7 @@ class ListIterInstance(_ListPayloadMixin):
         self._builder = builder
         self._ty = iter_type
         self._iter = make_listiter_cls(iter_type)(context, builder, iter_val)
+        self._datamodel = context.data_model_manager[iter_type.yield_type]
 
     @classmethod
     def from_list(cls, context, builder, iter_type, list_val):

--- a/numba/targets/listobj.py
+++ b/numba/targets/listobj.py
@@ -29,6 +29,8 @@ def make_payload_cls(list_type):
     Return the Structure representation of the given *list_type*'s payload
     (an instance of types.List).
     """
+    # Note the payload is stored durably in memory, so we consider it
+    # data and not value.
     return cgutils.create_struct_proxy(types.ListPayload(list_type),
                                        kind='data')
 

--- a/numba/targets/listobj.py
+++ b/numba/targets/listobj.py
@@ -10,9 +10,8 @@ from llvmlite import ir
 from numba import types, cgutils, typing
 from numba.targets.imputils import (builtin, builtin_attr, implement,
                                     impl_attribute, impl_attribute_generic,
-                                    iternext_impl, struct_factory,
-                                    impl_ret_borrowed, impl_ret_new_ref,
-                                    impl_ret_untracked)
+                                    iternext_impl, impl_ret_borrowed,
+                                    impl_ret_new_ref, impl_ret_untracked)
 from numba.utils import cached_property
 from . import quicksort, slicing
 
@@ -335,7 +334,6 @@ def list_len(context, builder, sig, args):
     return inst.size
 
 
-@struct_factory(types.ListIter)
 def make_listiter_cls(iterator_type):
     """
     Return the Structure representation of the given *iterator_type* (an

--- a/numba/targets/npdatetime.py
+++ b/numba/targets/npdatetime.py
@@ -6,8 +6,7 @@ from llvmlite.llvmpy.core import Type, Constant
 import llvmlite.llvmpy.core as lc
 
 from numba import npdatetime, types, cgutils
-from numba.targets.imputils import (builtin, implement, type_factory,
-                                    impl_ret_untracked)
+from numba.targets.imputils import builtin, implement, impl_ret_untracked
 
 
 if not npdatetime.NPDATETIME_SUPPORTED:
@@ -19,15 +18,6 @@ DATETIME64 = TIMEDELTA64 = Type.int(64)
 NAT = Constant.int(TIMEDELTA64, npdatetime.NAT)
 
 TIMEDELTA_BINOP_SIG = (types.Kind(types.NPTimedelta),) * 2
-
-
-@type_factory(types.NPDatetime)
-def llvm_datetime_type(context, tp):
-    return DATETIME64
-
-@type_factory(types.NPTimedelta)
-def llvm_timedelta_type(context, tp):
-    return TIMEDELTA64
 
 
 def scale_by_constant(builder, val, factor):

--- a/numba/targets/rangeobj.py
+++ b/numba/targets/rangeobj.py
@@ -6,10 +6,9 @@ import llvmlite.llvmpy.core as lc
 
 from numba import types, cgutils
 from numba.targets.imputils import (builtin, implement, iterator_impl,
-                                    struct_factory, impl_ret_untracked)
+                                    impl_ret_untracked)
 
 
-@struct_factory(types.RangeIteratorType)
 def make_range_iterator(typ):
     """
     Return the Structure representation of the given *typ* (an

--- a/numba/targets/slicing.py
+++ b/numba/targets/slicing.py
@@ -7,9 +7,8 @@ from llvmlite import ir
 from numba import cgutils, types, typing
 from .imputils import (builtin, builtin_attr, implement,
                        impl_attribute, impl_attribute_generic,
-                       iternext_impl, struct_factory,
-                       impl_ret_borrowed, impl_ret_new_ref,
-                       impl_ret_untracked)
+                       iternext_impl, impl_ret_borrowed,
+                       impl_ret_new_ref, impl_ret_untracked)
 
 
 def fix_index(builder, idx, size):

--- a/numba/targets/tupleobj.py
+++ b/numba/targets/tupleobj.py
@@ -6,7 +6,7 @@ from llvmlite import ir
 import llvmlite.llvmpy.core as lc
 
 from .imputils import (builtin, builtin_attr, implement, impl_attribute,
-                       impl_attribute_generic, iternext_impl, struct_factory,
+                       impl_attribute_generic, iternext_impl,
                        impl_ret_borrowed, impl_ret_untracked)
 from .. import typing, types, cgutils
 
@@ -118,7 +118,6 @@ def namedtuple_getattr(context, builder, typ, value, attr):
 #------------------------------------------------------------------------------
 # Tuple iterators
 
-@struct_factory(types.UniTupleIter)
 def make_unituple_iter(tupiter):
     """
     Return the Structure representation of the given *tupiter* (an

--- a/numba/tests/test_lists.py
+++ b/numba/tests/test_lists.py
@@ -269,6 +269,15 @@ def identity_usecase(n):
     c = a[:]
     return (a is b), (a is not b), (a is c), (a is not c)
 
+def bool_list_usecase():
+    # Exercise getitem, setitem, iteration with bool values (issue #1373)
+    l = [False]
+    l[0] = True
+    x = False
+    for v in l:
+        x = x ^ v
+    return l, x
+
 
 class TestLists(MemoryLeakMixin, TestCase):
 
@@ -616,6 +625,12 @@ class TestLists(MemoryLeakMixin, TestCase):
         pyfunc = identity_usecase
         cfunc = jit(nopython=True)(pyfunc)
         self.assertPreciseEqual(cfunc(3), pyfunc(3))
+
+    def test_bool_list(self):
+        # Check lists of bools compile and run successfully
+        pyfunc = bool_list_usecase
+        cfunc = jit(nopython=True)(pyfunc)
+        self.assertPreciseEqual(cfunc(), pyfunc())
 
 
 if __name__ == '__main__':

--- a/numba/tests/test_typeinfer.py
+++ b/numba/tests/test_typeinfer.py
@@ -121,12 +121,13 @@ class TestUnify(unittest.TestCase):
     def assert_unify(self, aty, bty, expected):
         ctx = typing.Context()
         template = "{0}, {1} -> {2} != {3}"
-        unified = ctx.unify_types(aty, bty)
-        self.assertEqual(unified, expected,
-                         msg=template.format(aty, bty, unified, expected))
-        unified = ctx.unify_types(bty, aty)
-        self.assertEqual(unified, expected,
-                         msg=template.format(bty, aty, unified, expected))
+        for unify_func in ctx.unify_types, ctx.unify_pairs:
+            unified = unify_func(aty, bty)
+            self.assertEqual(unified, expected,
+                             msg=template.format(aty, bty, unified, expected))
+            unified = unify_func(bty, aty)
+            self.assertEqual(unified, expected,
+                             msg=template.format(bty, aty, unified, expected))
 
     def assert_unify_failure(self, aty, bty):
         self.assert_unify(aty, bty, types.pyobject)
@@ -141,6 +142,16 @@ class TestUnify(unittest.TestCase):
             except KeyError:
                 expected = self.int_unify[key[::-1]]
             self.assert_unify(aty, bty, getattr(types, expected))
+
+    def test_bool(self):
+        aty = types.boolean
+        for bty in types.integer_domain:
+            self.assert_unify(aty, bty, bty)
+        # Not sure about this one, but it respects transitivity
+        cty = types.float64
+        self.assert_unify(aty, cty, cty)
+        for cty in types.real_domain:
+            self.assert_unify(aty, cty, cty)
 
     def unify_number_pair_test(self, n):
         """

--- a/numba/tests/test_typeinfer.py
+++ b/numba/tests/test_typeinfer.py
@@ -148,8 +148,6 @@ class TestUnify(unittest.TestCase):
         for bty in types.integer_domain:
             self.assert_unify(aty, bty, bty)
         # Not sure about this one, but it respects transitivity
-        cty = types.float64
-        self.assert_unify(aty, cty, cty)
         for cty in types.real_domain:
             self.assert_unify(aty, cty, cty)
 

--- a/numba/typeconv/rules.py
+++ b/numba/typeconv/rules.py
@@ -31,6 +31,7 @@ def _init_casting_rules(tm):
     tcr.safe_unsafe(types.uint16, types.int32)
     tcr.safe_unsafe(types.uint32, types.int64)
 
+    tcr.safe_unsafe(types.int16, types.float32)
     tcr.safe_unsafe(types.int32, types.float64)
 
     tcr.unsafe_unsafe(types.int32, types.float32)

--- a/numba/typing/context.py
+++ b/numba/typing/context.py
@@ -385,9 +385,9 @@ class BaseContext(object):
         # pairwise unification (with thanks to aterrel).
         def keyfunc(obj):
             """Uses bitwidth to order numeric-types.
-            Fallback to hash() for arbitary ordering.
+            Fallback to stable, deterministic sort.
             """
-            return getattr(obj, 'bitwidth', hash(obj))
+            return getattr(obj, 'bitwidth', 0)
 
         typelist = sorted(typelist, key=keyfunc)
         unified = typelist[0]
@@ -400,7 +400,7 @@ class BaseContext(object):
     def unify_pairs(self, first, second):
         """
         Try to unify the two given types.  A third type is returned,
-        or None in case of failure.
+        or pyobject in case of failure.
         """
         if first == second:
             return first
@@ -422,8 +422,15 @@ class BaseContext(object):
         # Other types with simple conversion rules
         conv = self.can_convert(fromty=first, toty=second)
         if conv is not None and conv <= Conversion.safe:
-            return conv
+            # Can convert from first to second
+            return second
 
+        conv = self.can_convert(fromty=second, toty=first)
+        if conv is not None and conv <= Conversion.safe:
+            # Can convert from second to first
+            return first
+
+        # Cannot unify
         return types.pyobject
 
 

